File: sdks/python/examples.py
"""Examples of using the OpenFireblocks Python SDK."""

from openfireblocks import Client, CreateKeyPairRequest, SigningRequest


def example_create_bitcoin_key():
    """Example: Create a Bitcoin threshold key."""
    client = Client(
        base_url="https://api.openfireblocks.io",
        api_key="your-api-key",
    )

    # Create a 4-of-7 threshold ECDSA key for Bitcoin
    key_pair = client.create_key_pair(
        CreateKeyPairRequest(
            name="bitcoin-cold-wallet",
            blockchain="bitcoin",
            threshold=4,
            total_parties=7,
        )
    )

    print(f"Created key pair: {key_pair.id}")
    print(f"Address: {key_pair.address}")
    print(f"Status: {key_pair.status}")
    print(f"Public Key: {key_pair.public_key}")

    return key_pair


def example_list_keys():
    """Example: List all key pairs."""
    client = Client(
        base_url="https://api.openfireblocks.io",
        api_key="your-api-key",
    )

    keys = client.list_key_pairs()

    print(f"Found {len(keys)} key pairs:")
    for key in keys:
        print(f"  - {key.id} ({key.blockchain}): {key.status}")


def example_sign_bitcoin_transaction():
    """Example: Sign a Bitcoin transaction."""
    client = Client(
        base_url="https://api.openfireblocks.io",
        api_key="your-api-key",
    )

    # Start signing
    sig_request = client.sign(
        SigningRequest(
            key_pair_id="key-pair-id",
            transaction="020000000117cc4c39f2a4b11d3d9e7f8e4c3b2a1f...",  # Bitcoin tx hex
        )
    )

    print(f"Signing request ID: {sig_request.id}")
    print(f"Status: {sig_request.status}")

    # Wait for completion
    try:
        completed = client.wait_for_signing(
            sig_request.id,
            max_wait=300,  # Wait up to 5 minutes
            poll_interval=2.0,  # Poll every 2 seconds
        )

        if completed.status == "completed":
            print(f"Signed successfully!")
            print(f"Signature: {completed.signature}")
            print(f"Signed Transaction: {completed.signed_transaction}")
            print(f"Latency: {completed.latency_ms}ms")
        else:
            print(f"Signing failed: {completed.error}")

    except TimeoutError as e:
        print(f"Timeout: {e}")


def example_sign_ethereum_transaction():
    """Example: Sign an Ethereum transaction."""
    client = Client(
        base_url="https://api.openfireblocks.io",
        api_key="your-api-key",
    )

    # Create Ethereum key
    eth_key = client.create_key_pair(
        CreateKeyPairRequest(
            name="ethereum-hot-wallet",
            blockchain="ethereum",
            threshold=3,
            total_parties=5,
        )
    )

    print(f"Created Ethereum key: {eth_key.address}")

    # Sign transaction
    sig_request = client.sign(
        SigningRequest(
            key_pair_id=eth_key.id,
            transaction="02f86a0102848c51d60085...",  # Ethereum tx hex
        )
    )

    print(f"Signing request: {sig_request.id} - {sig_request.status}")

    # Check status
    status = client.get_signing_status(sig_request.id)
    print(f"Current status: {status.status}")


def example_health_check():
    """Example: Check API health."""
    client = Client(
        base_url="https://api.openfireblocks.io",
        api_key="your-api-key",
    )

    health = client.health()
    print(f"API Status: {health.status}")
    print(f"Version: {health.version}")
    print(f"Timestamp: {health.timestamp}")


def example_multi_chain_management():
    """Example: Manage keys across multiple blockchains."""
    client = Client(
        base_url="https://api.openfireblocks.io",
        api_key="your-api-key",
    )

    blockchains = ["bitcoin", "ethereum", "solana", "cosmos"]
    keys = {}

    # Create keys for each blockchain
    for blockchain in blockchains:
        key = client.create_key_pair(
            CreateKeyPairRequest(
                name=f"{blockchain}-threshold-key",
                blockchain=blockchain,
                threshold=4,
                total_parties=7,
            )
        )
        keys[blockchain] = key
        print(f"Created {blockchain} key: {key.address}")

    # List all keys
    all_keys = client.list_key_pairs()
    print(f"\nTotal keys: {len(all_keys)}")


def example_idempotent_signing():
    """Example: Idempotent signing with automatic retry."""
    client = Client(
        base_url="https://api.openfireblocks.io",
        api_key="your-api-key",
    )

    # Create a signing request with explicit idempotency key
    idempotency_key = "my-unique-request-id-12345"

    request = SigningRequest(
        key_pair_id="key-pair-id",
        transaction="02f86a0102848c51d60085...",
        idempotency_key=idempotency_key,
    )

    # Submit the request
    sig_response = client.sign(request)
    print(f"Submitted: {sig_response.id}")

    # If the request fails and we retry, the API returns the same result
    # (idempotency ensures no double-signing)
    sig_response_retry = client.sign(request)
    assert sig_response.id == sig_response_retry.id
    print(f"Retry returned same request: {sig_response_retry.id}")


if __name__ == "__main__":
    print("Example 1: Create Bitcoin Key")
    print("-" * 50)
    # example_create_bitcoin_key()

    print("\n\nExample 2: List Keys")
    print("-" * 50)
    # example_list_keys()

    print("\n\nExample 3: Sign Bitcoin Transaction")
    print("-" * 50)
    # example_sign_bitcoin_transaction()

    print("\n\nExample 4: Health Check")
    print("-" * 50)
    # example_health_check()

    print("\nSet API key and base URL before running examples")

File: sdks/javascript/examples.js
/**
 * OpenFireblocks JavaScript SDK Examples
 */

const OpenFireblocksClient = require('./client');

const client = new OpenFireblocksClient(process.env.OPENFIREBLOCKS_API_KEY);

// Example 1: Create a new key pair for Bitcoin
async function example1_createBitcoinKey() {
  console.log('Creating Bitcoin key pair...');
  
  const ceremony = await client.createKey({
    blockchain: 'bitcoin',
    threshold: 4,
    total_parties: 7,
    name: 'Bitcoin Treasury Key',
  });
  
  console.log(`DKG Ceremony initiated: ${ceremony.id}`);
  console.log(`Status: ${ceremony.status}`);
  console.log(`Key Pair ID: ${ceremony.key_pair_id}`);
}

// Example 2: Sign a Bitcoin transaction
async function example2_signBitcoinTx() {
  console.log('Signing Bitcoin transaction...');
  
  const keyId = 'key_uuid_here';
  const txHex = '02000000...'; // Bitcoin transaction
  
  const signing = await client.sign({
    key_pair_id: keyId,
    transaction: txHex,
  });
  
  console.log(`Signing initiated: ${signing.id}`);
  
  // Wait for signature
  const completed = await client.waitForSignature(signing.id);
  console.log(`Signed transaction: ${completed.signed_transaction}`);
  console.log(`Latency: ${completed.latency_ms}ms`);
}

// Example 3: Sign an Ethereum transaction
async function example3_signEthereumTx() {
  console.log('Signing Ethereum transaction...');
  
  const keyId = 'key_uuid_here';
  const txData = {
    to: '0x...',
    value: '1000000000000000000', // 1 ETH
    data: '0x...',
    gasLimit: 21000,
    gasPrice: '20000000000',
    nonce: 0,
  };
  
  const signing = await client.sign({
    key_pair_id: keyId,
    transaction: JSON.stringify(txData),
  });
  
  const result = await client.waitForSignature(signing.id, { maxWaitTime: 120000 });
  console.log(`Signed transaction: ${result.signed_transaction}`);
}

// Example 4: List all key pairs
async function example4_listKeys() {
  console.log('Listing all key pairs...');
  
  const response = await client.listKeys({
    limit: 100,
    blockchain: 'ethereum',
  });
  
  console.log(`Found ${response.total} key pairs`);
  response.data.forEach(key => {
    console.log(`- ${key.name} (${key.address})`);
  });
}

// Example 5: Register customer and submit KYC
async function example5_customerKYC() {
  console.log('Registering customer...');
  
  const customer = await client.registerCustomer({
    name: 'Acme Corporation',
    email: 'legal@acme.io',
    country: 'US',
  });
  
  console.log(`Customer registered: ${customer.id}`);
  console.log(`KYC Status: ${customer.kyc_status}`);
  
  // Submit KYC document
  const kyc = await client.submitKYC(
    customer.id,
    '/path/to/passport.pdf',
    'passport'
  );
  
  console.log(`KYC submitted: ${kyc.verification_id}`);
}

// Example 6: Idempotent signing (retry-safe)
async function example6_idempotentSigning() {
  const idempotencyKey = 'merchant_order_12345';
  
  try {
    const signing = await client.sign({
      key_pair_id: 'key_uuid',
      transaction: 'tx_hex',
      idempotency_key: idempotencyKey,
    });
    
    const result = await client.waitForSignature(signing.id);
    console.log(`Signed with idempotency key: ${idempotencyKey}`);
  } catch (error) {
    // Retry with same idempotency key - will return same result
    console.error(`First attempt failed: ${error.message}`);
  }
}

// Example 7: Health check
async function example7_healthCheck() {
  const health = await client.health();
  console.log(`Service status: ${health.status}`);
  console.log(`Uptime: ${health.uptime_seconds}s`);
}

// Run examples
(async () => {
  try {
    await example1_createBitcoinKey();
    // await example2_signBitcoinTx();
    // await example3_signEthereumTx();
    // await example4_listKeys();
    // await example5_customerKYC();
    // await example6_idempotentSigning();
    // await example7_healthCheck();
  } catch (error) {
    console.error(`Error: ${error.message}`);
    process.exit(1);
  }
})();
